### ISSUE TITLE
Feature Request: punishment system

### ISSUE THREAD

**Issue by dotKuro** [yeldiRium/telegram-bots#35]
Since some people can't control oneself the bot should have a punishment system to keep leeting fun for everybody.

One who interrupts leet should be punished by being muted x days at leet time. x should
depend on how many times one already interrupted leet before and should be lower if one
participates leet with out interrupting regularly.

All this data should not be lost once someone disables and re-enables the bot.

**Comment by meisterrados** [comment]
The last request should be an independent one, since it would fix even more problems than just keeping state of punished users.
At this point, the highscore of 1337 participants is lost, if the bots is restarted.

**Comment by yeldiRium** [comment]
> At this point, the highscore of 1337 participants is lost, if the bots is restarted.

Slight correction: The highscore is lost if the bot is _re-enabled_. Restarting the bot as a service preserves the state.

But I do concur with both of your sentiments. I'll use this issue for the punishment feature request and open a new one for the persistence of data when disabling the bot. The latter will have to be implemented first.

> be punished by being muted x days at leet time. x should
depend on how many times one already interrupted leet before and should be lower if one
participates leet without interrupting regularly.

We will have to work out a system to calculate x.
For now I'll propose this:

Variables to tweak:

d := Number of days to take into consideration (should probably be infinite since cooldowns would allow people to just sometimes participate and annoy people without the punishment increasing)
s := severity factor

Data from context:

v := Number of times the asshole has interrupted leet in the last `d` days previously to the current interruption (so probably ever). I.e. if I interrupt leet and have never done so before, `v` is 0.
c := "Credit". Number of times the asshole has successfully participated in leeting in the past.

Initial punishment:

x := Number of muted days = `floor(2^(v*s/c))`

Examples:

```
s := 10
v := 4
c := 25
x = floor(2^(4*10/25)) = floor(2^(2/5)) = 3
```

```
s := 10
v := 2
c := 30
x = floor(2^(2*10/25)) = 1
```

```
s := 10
v := 2
c := 10
x = floor(2^(2*5/10)) = 4
```

This seems promising at first glance but we'll probably have to play around with `s`. Also I have no idea if the `v`s and `c`s I've chosen are realistic. We unfortunately have no statistics at the moment.

**Comment by meisterrados** [comment]
Let me drop another example for the proposed formular:
Say, I have never interrupted 1337 time.
Then, v is 0.
s and c don't play a role to point out what the problem is. Let's check the formular:
x = floor( 2 ^ (0*s/c)) = 1.
Consequently, everyone is muted from the chat.
Quick fix would be:
x = floor (2 ^ (v * s / c)) - 1

**Comment by yeldiRium** [comment]
You are correct but your fix is not necessary. `v` is the number of interruptions _before_ the current incident. The resulting 1 is intended for first time offenders.

**Comment by meisterrados** [comment]
Okay, so that's my error, then.
So everyone being the asshole for the first time will result in an un-warned mute. I'd give anyone for the first a message explaining the issue.

**Comment by yeldiRium** [comment]
> I'd give anyone for the first a message explaining the issue.

That's a good point. Fair warning for first time offenders. I still wouldn't subtract a flat amount, since the formula currently scales rather slowly in the beginning. I'd check explicitly for first offense.

**Comment by meisterrados** [comment]
We should be careful with dividing by c, which is 0 if someone fails 1337 and has never before participated correctly.
Here's another calculation. Say, somebody fails twice in a row, having participated correctly only once (I'd rather insert 0 for my example but that'd make mathematicians sad), we have:
x = (2 ^ (1 * 10 / 1)) = 1024 days of muting.
We should rather not use an exponential function, or try to find a better suited factor s.

**Comment by yeldiRium** [comment]
Again a good point. So far i just assumed that the number of successful participations should be a bit higher since 13:37 is every day. But you are right - a beginners mistake could easily exclude someone for figuratively ever. I'll think about this for a bit and come back to it.

Regarding `c=0` being problematic: I have no issue with just adding one to c. That would at least solve the problem eith dividing by zero.

**Comment by strangedev** [comment]
I've come up with a small simulation based on one score per user. If the user surpasses a certain threshold, they're muted during leet-time (or similar punishment). Bad actions increase the score and the score naturally decays back to 0 after enough time (depending on how high the score was before).

Take a look: https://github.com/strangedev/banhammer-the-simulation
